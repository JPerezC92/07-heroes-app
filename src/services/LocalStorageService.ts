class LocalStorageService {
  static get<T>(key: string): T | null {
    const value = localStorage.getItem(key);

    if (value) return JSON.parse(value);
    return null;
  }

  static save<T>(key: string, value: T) {
    localStorage.setItem(key, JSON.stringify(value));
  }
}

export default LocalStorageService;
