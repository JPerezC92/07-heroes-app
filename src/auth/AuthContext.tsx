import React, { createContext, useContext, useEffect, useReducer } from "react";
import { AuthAction, authReducer } from "./AuthReducer";

export interface IUser {
  logged: boolean;
  name: string | null;
}

interface IAuthContext extends IUser {
  dispatch: React.Dispatch<AuthAction>;
}

export const initialAuthContextState = {
  logged: false,
  name: null,
} as IAuthContext;

const AuthContext = createContext(initialAuthContextState);

const useAuthState = () => {
  const context = useContext(AuthContext);
  return context;
};

const init = (initialState: IUser) => {
  const userStorageString = localStorage.getItem("user");

  if (userStorageString) return JSON.parse(userStorageString) as IUser;

  return initialState;
};

const AuthContextProvider: React.FC = ({ children }) => {
  const [user, dispatch] = useReducer(
    authReducer,
    initialAuthContextState,
    init
  );

  const value: IAuthContext = { ...user, dispatch };

  useEffect(() => localStorage.setItem("user", JSON.stringify(user)), [user]);

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export { AuthContextProvider, useAuthState };
