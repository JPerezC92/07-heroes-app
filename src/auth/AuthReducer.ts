import { IAuthContext } from "./AuthContext";

enum ActionType {
  login = "[auth] login",
  logout = "[auth] logout",
}

export const authReducer = (
  state: IAuthContext,
  action: { type: ActionType; payload: ObjectConstructor }
): IAuthContext => {
  switch (action.type) {
    case ActionType.login:
      return { ...action.payload, logged: true };

    case ActionType.logout:
      return { logged: false };

    default:
      return state;
  }
};
