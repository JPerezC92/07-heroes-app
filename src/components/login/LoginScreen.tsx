import { RouteComponentProps } from "react-router-dom";
import { useAuthState } from "../../auth/AuthContext";
import { AuthActionType } from "../../auth/AuthReducer";
import LocalStorageService from "../../services/LocalStorageService";

const LoginScreen = ({ history }: RouteComponentProps) => {
  const { dispatch } = useAuthState();

  const lastpath = LocalStorageService.get<string>("lastpath") || "/";

  const handleLogin = () => {
    history.replace(lastpath);

    dispatch({
      type: AuthActionType.login,
      payload: { name: "Philip" },
    });
  };

  return (
    <div className="container mt-5">
      <h1>Login Screen</h1>
      <hr />
      <button className="btn btn-primary" onClick={handleLogin}>
        Login
      </button>
    </div>
  );
};

export default LoginScreen;
