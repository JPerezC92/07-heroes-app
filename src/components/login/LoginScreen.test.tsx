import { mount } from "enzyme";
import { MemoryRouter, RouteComponentProps } from "react-router-dom";
import { routeComponentPropsFixture } from "../../test/fixtures/routeComponentPropsFixture";
import { AuthActionType } from "../../auth/AuthReducer";
import LoginScreen from "./LoginScreen";
import { useAuthState } from "../../auth/AuthContext";
import LocalStorageService from "../../services/LocalStorageService";

jest.mock("../../auth/AuthContext", () => {
  return { useAuthState: jest.fn() };
});

describe("Test on <LoginScreen />", () => {
  const loginScreenProps: RouteComponentProps = {
    ...routeComponentPropsFixture,
    history: { ...routeComponentPropsFixture.history, replace: jest.fn() },
  };

  const useAuth = {
    user: { logged: true, name: null },
    dispatch: jest.fn(),
  };
  (useAuthState as jest.MockedFunction<any>).mockReturnValue(useAuth);

  const wrapper = mount(
    <MemoryRouter>
      <LoginScreen {...loginScreenProps} />
    </MemoryRouter>
  );

  test("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should realize a dispatch and navigation", () => {
    const button = wrapper.find("button");

    button.simulate("click");
    expect(loginScreenProps.history.replace).toHaveBeenCalledWith("/");
    expect(useAuth.dispatch).toHaveBeenCalledWith({
      payload: { name: "Philip" },
      type: AuthActionType.login,
    });
  });

  test("should rnavigate to /dc", () => {
    const useAuth = {
      user: { logged: true, name: null },
      dispatch: jest.fn(),
    };
    (useAuthState as jest.MockedFunction<any>).mockReturnValue(useAuth);

    LocalStorageService.save("lastpath", "/dc");

    const wrapper = mount(
      <MemoryRouter>
        <LoginScreen {...loginScreenProps} />
      </MemoryRouter>
    );

    wrapper.find("button").simulate("click");
    expect(loginScreenProps.history.replace).toHaveBeenCalledWith("/dc");
  });
});
