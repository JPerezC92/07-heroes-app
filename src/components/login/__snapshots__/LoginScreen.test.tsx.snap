// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <LoginScreen /> should render correctly 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login Screen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
